### PR TITLE
Wangc/convert pp to interface

### LINKED ISSUE #1529: [TASK] Restructure PostProcessor based BaseEntity and BaseInterface Structure.
--------
Issue Description
--------
**Is your feature request related to a problem? Please describe.**
Recent PR #1491 restructured and created two base classes, i.e. BaseEntity and BaseInterface. Now everything in RAVEN should now inherit from either BaseEntity or BaseInterface, not BaseType. The first step for this conversion is to convert PostProcessors to use this structure. 

See also issue #1486 regarding BaseEntity and BaseInterace, and PR #1491 for the design.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Restructure PostProcessor Entity and PostProcessor Interface. Let PostProcessor Entity inherit from Model, and PostProcessor Interface inherit form BaseInterface. 
See #1486  

Close #1529 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  @abc.abstractmethod
+  def run(self, input):
+    """
+      This method executes the postprocessor action.
+      @ In,  input, object, object containing the data to process.
+      Should avoid to use (inputToInternal output), and passing xarray directly/dataset
+      Possible inputs include: dict, xarray.Dataset, pd.DataFrame
+      @ Out, dict, xarray.Dataset, pd.DataFrame --> I think we can avoid collectoutput in the plugin pp
+    """
+    pass
```
Comment (PaulTalbot-INL) [inline framework/Models/PostProcessors/PostProcessorInterface.py:101]: no pass on abstract methods

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      @ Out, None
+    """
+    super()._handleInput(paramInput)
+
+  def initialize(self, runInfo, inputs, initDict=None):
+    """
+      Method to initialize the PostProcessor
+      @ In, runInfo, dict, it is the run info from the jobHandler
+      @ In, inputs, list, it is a list containing whatever is passed with an input role in the step
+      @ In, initDict, dict, optional, dictionary of all objects available in the step is using this model
```
Comment (PaulTalbot-INL) [inline framework/Models/PostProcessors/PostProcessorInterface.py:83]: no `@Out`

Also, does `initDict` need to be part of the arguments? It doesn't seem to get used. Is this because it's in the argument list of the other postprocessors?

**Review by PaulTalbot-INL — CHANGES_REQUESTED** [review]
A couple small comments, but this looks good otherwise. I'm sure this was tedious to convert; thanks for your effort!

File: framework/BaseClasses/BaseInterface.py [hunk 3]
```diff
@@ -97,16 +98,14 @@ def provideExpectedMetaKeys(self):
     """
     return self.metadataKeys, self.metadataParams
 
-  def addMetaKeys(self, args, params=None):
+  def addMetaKeys(self,args, params={}):
```
[the remark this answers is not in the document]
Comment (wangcj05) [reply framework/BaseClasses/BaseInterface.py]: Good to know this. Thanks and added.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 1 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py]: removed

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 2 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:83]: added. The "initDict" is used by LimitSurface PP.

**Comment by wangcj05** [comment]
@PaulTalbot-INL I have resolved your comments. Thanks.

**Review by PaulTalbot-INL — DISMISSED** [review]
Code changes approved. Continuing with the checklist ...

**Comment by PaulTalbot-INL** [comment]
With the HPC down this week, we're stuck waiting on the tests unless @alfoa green lights an alternate testing strategy. I only have access to OSX myself right now.

File: framework/Models/PostProcessor.py [hunk 4]
```diff
@@ -99,34 +105,54 @@ def specializeValidateDict(cls):
...
+
+  def _readMoreXML(self,xmlNode):
+    """
+      Function to read the portion of the xml input that belongs to this specialized class
+      and initialize some stuff based on the inputs got
+      @ In, xmlNode, xml.etree.ElementTree.Element, Xml element node
       @ Out, None
     """
-    Model._handleInput(self, paramInput)
+    Model._readMoreXML(self, xmlNode)
```
Comment (alfoa) [inline framework/Models/PostProcessor.py:127]: why the readMoreXML?

Comment (alfoa) [reply framework/Models/PostProcessor.py:127]: this should disappear

File: framework/Models/PostProcessor.py [hunk 5]
```diff
@@ -188,10 +191,10 @@ def evaluateSample(self, myInput, samplerType, kwargs):
           the second item will be the output of this model given the specified
           inputs
     """
-    Input = self.createNewInput(myInput,samplerType, **kwargs)
-    if Input is not None and len(Input) == 0:
-      Input = None
-    returnValue = (Input, self.run(Input))
+    ppInput = self.createNewInput(myInput,samplerType, **kwargs)
```
Comment (alfoa) [inline framework/Models/PostProcessor.py:194]: I am conflicted here again. the samplerType is set at the Step stage...so for PostProcessor steps it will always be None. I dnk but I think it is not needed and it can cause confusion for the PostProcessor developer

Comment (alfoa) [reply framework/Models/PostProcessor.py:194]: I know it was there arleady, but probably it is a good time to remove it

File: framework/Models/PostProcessor.py [hunk 6]
```diff
@@ -207,9 +210,9 @@ def submit(self,myInput,samplerType,jobHandler,**kwargs):
         @ Out, None
     """
     kwargs['forceThreads'] = True
-    Model.submit(self,myInput, samplerType, jobHandler,**kwargs)
+    super().submit(myInput, samplerType, jobHandler,**kwargs)
```
Comment (alfoa) [inline framework/Models/PostProcessor.py:213]: as above

File: framework/Models/PostProcessor.py [hunk 7]
```diff
@@ -218,31 +221,26 @@ def collectOutput(self,finishedJob,output,options=None):
...
+     """
+     # get keys as per base class
+     metaKeys,metaParams = super().provideExpectedMetaKeys()
+     # add postprocessor keys
+     try:
+       keys, params = self._pp.provideExpectedMetaKeys()
+       metaKeys = metaKeys.union(keys)
+       metaParams.update(params)
+     except AttributeError:
+       pass # either "interface" has no method for returning meta keys, or "interface" is not established yet.
```
Comment (alfoa) [inline framework/Models/PostProcessor.py:245]: this should not be possible. If it is in the base class, this AttributeError should not ever happen.

File: framework/Models/PostProcessors/CrossValidation.py [hunk 8]
```diff
@@ -329,7 +327,7 @@ def run(self, inputIn):
           scoreDict[varName] = np.atleast_1d(np.mean(np.atleast_1d(metricValues)))
       return scoreDict
 
-  def collectOutput(self,finishedJob, output):
+  def collectOutput(self,finishedJob, output, options=None):
```
Comment (alfoa) [inline framework/Models/PostProcessors/CrossValidation.py:330]: options for?

File: framework/Models/PostProcessors/DataClassifier.py [hunk 9]
```diff
@@ -40,7 +40,7 @@ class cls.
       @ Out, inputSpecification, InputData.ParameterInput, class to use for
         specifying input of cls.
     """
-    inputSpecification = super(DataClassifier, cls).getInputSpecification()
```
Comment (alfoa) [inline framework/Models/PostProcessors/DataClassifier.py:43]: please open an issue for all the PP to add description of the parameters. We need to expand the automatic documentation coverage

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+      @ In, runInfoDict, dict, the dictionary containing the runInfo (read in the XML input file)
+      @ Out, None
+    """
+    super().__init__()
+    self.inputCheckInfo  = []     # List of tuple, i.e input objects info [('name','type')]
+    self.action = None            # action
+    self.workingDir = ''          # path for working directory
+    self.printTag = 'PostProcessorInterface'
+    self.outputDataset  = False # True if the user wants to dump the outputs to dataset
+    self.validDataType = ['PointSet','HistorySet'] # The list of accepted types of DataObject
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:61]: should not be this added in the Models.py validation? in addition,if we have those here why in the other PR we still check for "Databases"?

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    ## Currently, we have used both DataObject.addRealization and DataObject.load to
+    ## collect the PostProcessor returned outputs. DataObject.addRealization is used to
+    ## collect single realization, while DataObject.load is used to collect multiple realizations
+    ## However, the DataObject.load can not be directly used to collect single realization
+    ## One possible solution is all postpocessors return a list of realizations, and we only
+    ## use addRealization method to add the collections into the DataObjects
+    self.outputMultipleRealizations = False
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:68]: hm, this should be fixed. The list of realizations might be slow. Dataset?

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  def collectOutput(self,finishedJob, output, options=None):
+    """
+      Method that collects the outputs from the "run" method of the PostProcessor
+      @ In, finishedJob, InternalRunner object, instance of the run just finished
+      @ In, output, "DataObjects" object, output where the results of the calculation needs to be stored
+      @ In, options, dict, optional, not used in PostProcessor.
+        dictionary of options that can be passed in when the collect of the output is performed by another model (e.g. EnsembleModel)
+      @ Out, None
+    """
+    if output.type not in self.validDataType:
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:112]: this test should be done in the Models.py validate dict. In addition, this should be done at least at the initialize, otherwise we error out at the end of the calculation (that can be very long for large data to process)

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+      self.raiseAnError(IOError, 'Output type', str(output.type), 'is not allowed!')
+    evaluation = finishedJob.getEvaluation()
+    if isinstance(evaluation, Runners.Error):
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:115]: please open an issue. This "error" must be handled at the JobHandler level. We should not reach this point.

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    if output.type in ['PointSet','HistorySet']:
+      if self.outputDataset:
+        self.raiseAnError(IOError, "DataSet output is required, but the provided type of DataObject is", output.type)
+      self.raiseADebug('Dumping output in data object named ' + output.name)
+      if self.outputMultipleRealizations:
+        if 'dims' in outputRealization:
+          dims = outputRealization['dims']
+        else:
+          dims = {}
+        print(outputRealization.keys())
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:128]: remove

File: framework/Models/PostProcessors/PostProcessorInterface.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        output.load(outputRealization['data'], style='dict', dims=dims)
```
Comment (alfoa) [inline framework/Models/PostProcessors/PostProcessorInterface.py:129]: the load should be modified to also accept a single realizaiton.

PS. I think that the PP must return a dataset. dictionaries are slow

File: framework/Models/PostProcessors/__init__.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+from .EconomicRatio import EconomicRatio
+## These utilize the optional prequisite library PySide, so don't error if they
+## do not import appropriately.
+additionalModules = []
+try:
+  from .TopologicalDecomposition import QTopologicalDecomposition
+  from .DataMining import QDataMining
+  additionalModules.append(QTopologicalDecomposition)
+  additionalModules.append(QDataMining)
+except ImportError:
```
Comment (alfoa) [inline framework/Models/PostProcessors/__init__.py:51]: please see PR #1503. These double imports are not needed. The only imports needed are in the Factory.py (it was supposed to be addressed in the PR #1480)

**Review by alfoa — CHANGES_REQUESTED** [review]

File: framework/Models/PostProcessor.py (hunk 4 above)
Comment (PaulTalbot-INL) [reply framework/Models/PostProcessor.py:127]: there was a lot of talk about this in the Entity/Interface PR; unfortunately, the conversion to purely using `handleInput` is more than just adding `handleInput`; the `Simulation` is set up to only purely use `handleInput` if a particular flag is thrown by the module (or factory), which requires the module (or factory) to create the input specs for the object. This only exists for a very few objects in RAVEN (the Database I think is the only one I recall offhand).

In short, some significant work needs to be done to eliminate `readMoreXML` given how `Simulation` interacts with the modules.

Comment (wangcj05) [reply framework/Models/PostProcessor.py:127]: This is needed to handle Assembler correctly. I think we really need to clean up our code to remove readMoreXML, otherwise, you will see this legacy method called in several places.

File: framework/Models/PostProcessor.py (hunk 5 above)
Comment (wangcj05) [reply framework/Models/PostProcessor.py:194]: I agree to remove it.

File: framework/Models/PostProcessors/CrossValidation.py (hunk 8 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/CrossValidation.py:330]: This options is from Model entity that you added. I keep it here for consistency. We can remove it if you agree.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 10 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:61]: This is for specific post-processors. They can accept only PointSet or both PointSet and HistorySet. I do not see the option that we can check at the Model level. It is PP-dependent options.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 11 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:68]: We use both realization and load method to collect the realizations for our PPs. It is PP-dependent, and can talk some time to convert them to use the same method. I keep this option, so we can split work to convert them after this both this PR and PR plugin get merged.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 12 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:112]: See my other comment about this. I agree to move it to initialize method.

File: framework/Models/PostProcessors/__init__.py (hunk 16 above)
Comment (PaulTalbot-INL) [reply framework/Models/PostProcessors/__init__.py:51]: I think that observation came after that PR, right? Either way, I concur, I like the idea that classes are only available through the factories. This may require some standardization so we can check types consistently; previously we would import the module and check instance, such as
```python
from Samplers import MonteCarlo
if isinstance(mySampler, MonteCarlo):
  ...
```
This might be harder with factories, but should be standardizable without checking against strings.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 15 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:129]: Let's have a discussion for this.

File: framework/Models/PostProcessor.py (hunk 5 above)
Comment (wangcj05) [reply framework/Models/PostProcessor.py]: I have updated the docstring for better clarification. I think we can keep "samplerType" which will consistent with other Models at entity level. In addition, these methods will only at entity level, and we will not expose this method to post-processor developer. In other words, the pp developers will not see these methods, so there is no confusion for them.

File: framework/Models/PostProcessor.py (hunk 6 above)
Comment (wangcj05) [reply framework/Models/PostProcessor.py:213]: See above comment. In addition, 'samplerType' is needed, since we are calling the method from Model class.

File: framework/Models/PostProcessor.py (hunk 7 above)
Comment (wangcj05) [reply framework/Models/PostProcessor.py]: updated. I have removed the try-except.

File: framework/Models/PostProcessors/CrossValidation.py (hunk 8 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/CrossValidation.py]: I have removed the options.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 12 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:110]: I have checked the initialize method, we only pass the input DataObjects, not the output DataObjects.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 14 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py]: removed.

File: framework/Models/PostProcessors/__init__.py (hunk 16 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/__init__.py]: I have cleaned up the __init__.py file

**Comment by wangcj05** [comment]
@alfoa I have resolved your comments.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 13 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:113]: see #1524

File: framework/Models/PostProcessors/DataClassifier.py (hunk 9 above)
Comment (wangcj05) [reply framework/Models/PostProcessors/DataClassifier.py:43]: see #1525

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 12 above)
Comment (alfoa) [reply framework/Models/PostProcessors/PostProcessorInterface.py:110]: In this:
```
  def specializeValidateDict(cls):
    """
      This method describes the types of input accepted with a certain role by the model class specialization
      @ In, None
      @ Out, None
    """
```

Comment (wangcj05) [reply framework/Models/PostProcessors/PostProcessorInterface.py:110]: That's for the entity class, not for the interface class. It seems to me we do not have the structure for the interface class yet.

File: framework/Models/PostProcessor.py (hunk 4 above)
Comment (alfoa) [reply framework/Models/PostProcessor.py]: Okay for now, this is a priority.

File: framework/Models/PostProcessor.py (hunk 5 above)
Comment (alfoa) [reply framework/Models/PostProcessor.py]: Please open under discussion issue.

File: framework/Models/PostProcessor.py (hunk 6 above)
Comment (alfoa) [reply framework/Models/PostProcessor.py:213]: See comment below. The samplerType should be encpasulated in the kwargs so usable by only the "entities" that need it.

File: framework/Models/PostProcessors/DataClassifier.py (hunk 9 above)
Comment (alfoa) [reply framework/Models/PostProcessors/DataClassifier.py:43]: Please update the issue including that the _readMoreXML must be removed once the process is complete

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 10 above)
Comment (alfoa) [reply framework/Models/PostProcessors/PostProcessorInterface.py:61]: see comment below...the "validateDict" approach can be expanded to the entity baseInterface classes. Right?

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 11 above)
Comment (alfoa) [reply framework/Models/PostProcessors/PostProcessorInterface.py:68]: ok. As long as we agree that this will need to disapaer . Same API for all.

File: framework/Models/PostProcessors/PostProcessorInterface.py (hunk 15 above)
Comment (alfoa) [reply framework/Models/PostProcessors/PostProcessorInterface.py:126]: we can, but the "gate" (if statement) should be included in the load method

**Review by alfoa — APPROVED** [review]

**Comment by alfoa** [comment]
Some regression test machines have connectivity issues currently under investigation by @joshua-cogliati-inl . The tests have been run locally and are green.

The PR is approved and can be merged.

**Comment by alfoa** [comment]
Checklist passed.

PR approved.

Merging
